### PR TITLE
null safety in ElevatedButton.icon

### PR DESCRIPTION
**Pull request by Abhishek01039**
## Description

Null safety label is there but it is not present in the constructor.
https://master-api.flutter.dev/flutter/material/ElevatedButton/ElevatedButton.icon.html

## Related Issues

*Replace this paragraph with a list of issues related to this PR from our [issue database]. Indicate, which of these issues are resolved or fixed by this PR. There should be at least one issue listed here.*

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by sidrao2006** [comment]
cc @goderbauer , @Piinks

**Review by Piinks — COMMENTED** [review]
Hi @Abhishek01039 Thanks for the contribution! Can you add a test that verifies these parameters can be set to null? That will prevent your change from being un-done in the future. Thanks!

File: packages/flutter/test/material/elevated_button_test.dart [hunk 1]
```diff
@@ -375,6 +375,82 @@ void main() {
     expect(didLongPressButton, isTrue);
   });
 
+   testWidgets('ElevatedButton.icon onPressed and onLongPress callbacks are correctly called when non-null', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/elevated_button_test.dart:378]: Hey @Abhishek01039  these don't appear to test the change you are making. I don't think you will need to test callbacks for your change. Just adding a test that sets the related parameters to null and validates their values should be sufficient, similar to this test: https://github.com/flutter/flutter/blob/e422d5c724309eef3b8a24e8d9bc28943b04cc27/packages/flutter/test/material/elevated_button_test.dart#L14

File: packages/flutter/test/material/elevated_button_test.dart [hunk 2]
```diff
@@ -375,6 +375,82 @@ void main() {
...
+    Finder elevatedButton;
+
+    Widget buildFrame({ VoidCallback? onPressed, VoidCallback? onLongPress }) {
+      return Directionality(
+        textDirection: TextDirection.ltr,
+        child: ElevatedButton.icon(
+          label: const Text('button'),
+          icon: const Icon(Icons.add),
+          onPressed: onPressed,
+          onLongPress: onLongPress,
```
Comment (Piinks) [inline packages/flutter/test/material/elevated_button_test.dart:389]: None of these are null, so this does not validate that they can be null.

File: packages/flutter/test/material/elevated_button_test.dart [hunk 3]
```diff
@@ -375,6 +375,82 @@ void main() {
...
+
+    // onPressed null, onLongPress null.
+    await tester.pumpWidget(
+      buildFrame(onPressed: null, onLongPress: null),
+    );
+    elevatedButton = find.byType(ElevatedButton);
+    expect(tester.widget<ElevatedButton>(elevatedButton).enabled, false);
+  });
+
+  testWidgets('ElevatedButton.icon onPressed and onLongPress callbacks are distinctly recognized', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/elevated_button_test.dart:422]: Again, this does not appear to test anything related to your change.

File: packages/flutter/lib/src/material/elevated_button.dart [hunk 4]
```diff
@@ -82,11 +82,11 @@ class ElevatedButton extends ButtonStyleButton {
...
-    VoidCallback onLongPress,
-    ButtonStyle style,
-    FocusNode focusNode,
+    Key? key,
+    required VoidCallback? onPressed,
+    VoidCallback? onLongPress,
+    ButtonStyle? style,
+    FocusNode? focusNode,
     bool autofocus,
     Clip clipBehavior,
```
Comment (sidrao2006) [inline packages/flutter/lib/src/material/elevated_button.dart:91]: I think `autofocus` and `clipBehaviour` should be nullable as well
https://github.com/flutter/flutter/blob/dde20448f7bce0b98da63a128488100b0bd495e5/packages/flutter/lib/src/material/elevated_button.dart#L377

Comment (Abhishek01039) [reply packages/flutter/lib/src/material/elevated_button.dart:91]: Yes, you are right. we need null safety here as well
I have done this.
Thanks

**Comment by sidrao2006** [comment]
Hey @Piinks , 
below is a passing test...is it enough?

```
  testWidgets('ElevatedButton.icon defaults', (WidgetTester tester) async {
  final ElevatedButton button = ElevatedButton.icon(
            onPressed: () {},
            icon: const Icon(Icons.elevator),
            label: const Text('Elevated button'),
          );
    await tester.pumpWidget(
      MaterialApp(
        home: Center(
          child: button,
        ),
      ),
    );

    // Finish the elevation animation, final background color change.
    await tester.pumpAndSettle();

    final Finder elevatedButtonWithIconFinder = find.byWidget(button);
    final ElevatedButton elevatedButtonWithIcon = tester.widget(elevatedButtonWithIconFinder);

    expect(elevatedButtonWithIcon.key, null);
    expect(elevatedButtonWithIcon.autofocus, false);
    expect(elevatedButtonWithIcon.clipBehavior, Clip.none);
    expect(elevatedButtonWithIcon.focusNode, null);
    expect(elevatedButtonWithIcon.style, null);
    expect(elevatedButtonWithIcon.onLongPress, null);
});
```

**Comment by Abhishek01039** [comment]
Thanks, @sidrao2006 for your time
I have done this so far @Piinks 
```
testWidgets('ElevatedButton.icon defaults', (WidgetTester tester) async {
    const ColorScheme colorScheme = ColorScheme.light();

    // Enabled ElevatedButton.icon
    await tester.pumpWidget(
      MaterialApp(
        theme: ThemeData.from(colorScheme: colorScheme),
        home: Center(
          child: ElevatedButton.icon(
            onPressed: () { },
            label: const Text('button'),
            icon: const Icon(Icons.add),
          ),
        ),
      ),
    );

    final Finder rawButtonMaterial = find.descendant(
      of: find.byType(ElevatedButton),
      matching: find.byType(Material),
    );


    Material material = tester.widget<Material>(rawButtonMaterial);
    expect(material.animationDuration, const Duration(milliseconds: 200));
    expect(material.borderOnForeground, true);
    expect(material.borderRadius, null);
    expect(material.clipBehavior, Clip.none);
    expect(material.color, colorScheme.primary);
    expect(material.elevation, 2);
    expect(material.shadowColor, const Color(0xff000000));
    expect(material.shape, RoundedRectangleBorder(borderRadius: BorderRadius.circular(4)));
    expect(material.textStyle!.color, colorScheme.onPrimary);
    expect(material.textStyle!.fontFamily, 'Roboto');
    expect(material.textStyle!.fontSize, 14);
    expect(material.textStyle!.fontWeight, FontWeight.w500);
    expect(material.type, MaterialType.button);

    final Offset center = tester.getCenter(find.byType(ElevatedButton));
    final TestGesture gesture = await tester.startGesture(center);
    await tester.pump(); // start the splash animation
    await tester.pump(const Duration(milliseconds: 100)); // splash is underway
    final RenderObject inkFeatures = tester.allRenderObjects.firstWhere((RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures');
    expect(inkFeatures, paints..circle(color: colorScheme.onPrimary.withAlpha(0x3d))); // splash color is onPrimary(0.24)

    // Only elevation changes when enabled and pressed.
    material = tester.widget<Material>(rawButtonMaterial);
    expect(material.animationDuration, const Duration(milliseconds: 200));
    expect(material.borderOnForeground, true);
    expect(material.borderRadius, null);
    expect(material.clipBehavior, Clip.none);
    expect(material.color, colorScheme.primary);
    expect(material.elevation, 8);
    expect(material.shadowColor, const Color(0xff000000));
    expect(material.shape, RoundedRectangleBorder(borderRadius: BorderRadius.circular(4)));
    expect(material.textStyle!.color, colorScheme.onPrimary);
    expect(material.textStyle!.fontFamily, 'Roboto');
    expect(material.textStyle!.fontSize, 14);
    expect(material.textStyle!.fontWeight, FontWeight.w500);
    expect(material.type, MaterialType.button);

    await gesture.up();
    await tester.pumpAndSettle();

    // Disabled ElevatedButton.icon
    await tester.pumpWidget(
      MaterialApp(
        theme: ThemeData.from(colorScheme: colorScheme),
        home: Center(
          child: ElevatedButton.icon(
            onPressed: null,
            label: const Text('button'),
            icon: const Icon(Icons.add),
          ),
        ),
      ),
    );

    // Finish the elevation animation, final background color change.
    await tester.pumpAndSettle();

    material = tester.widget<Material>(rawButtonMaterial);
    expect(material.animationDuration, const Duration(milliseconds: 200));
    expect(material.borderOnForeground, true);
    expect(material.borderRadius, null);
    expect(material.clipBehavior, Clip.none);
    expect(material.color, colorScheme.onSurface.withOpacity(0.12));
    expect(material.elevation, 0.0);
    expect(material.shadowColor, const Color(0xff000000));
    expect(material.shape, RoundedRectangleBorder(borderRadius: BorderRadius.circular(4)));
    expect(material.textStyle!.color, colorScheme.onSurface.withOpacity(0.38));
    expect(material.textStyle!.fontFamily, 'Roboto');
    expect(material.textStyle!.fontSize, 14);
    expect(material.textStyle!.fontWeight, FontWeight.w500);
    expect(material.type, MaterialType.button);
  });
```

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 4 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/material/elevated_button.dart:91]: ```Hmm, looks like autofocus and clipBehaviour can be nullable only in theory...```
I have removed null safety from clipBehaviour and autofocus.
Thanks

File: packages/flutter/test/material/elevated_button_test.dart (hunk 3 above)
Comment (Abhishek01039) [reply packages/flutter/test/material/elevated_button_test.dart:422]: ok @Piinks

**Review by Piinks — COMMENTED** [review]
@Abhishek01039 this change is making some parameters here nullable. The test should make sure no one un-does that change in the future. I believe something like pumping this widget with your nullable parameters set to null will suffice. If someone makes these parameters non-nullable again, the test will fail.
```dart
ElevatedButton.icon(
    // These are the parameters you are making nullable.
    key: null,
    onPressed: null,
    onLongPress: null,
    style: null,
    focusNode: null,
)
```

**Review by Piinks — APPROVED** [review]
LGTM, thanks @Abhishek01039!

**Comment by Abhishek01039** [comment]
Thanks @Piinks

**Comment by PixelToast** [comment]
This PR wasn't be merged automatically because #69596 landed on red before it, which implements the same changes.

Even though we couldn't get this landed in the end, thank you for taking the time to write this PR and adding tests!

**Comment by HansMuller** [comment]
@Abhishek01039 - sorry to have landed https://github.com/flutter/flutter/pull/69596 when you'd already been working on a solution for a week; I'm afraid I just overlooked this PR. Thanks for the contribution!

**Comment by Abhishek01039** [comment]
@HansMuller 
No issue with that.
Thanks

**Pull request closed without merging**